src/Bicep.Core/BicepCompiler.cs

```diff
@@ -69,27 +69,30 @@ public async Task<Compilation> CreateCompilation(Uri bicepUri, IReadOnlyWorkspac
         // however we still want to surface as many errors as we can for the module refs that are valid
         // so we will try to restore modules with valid refs and skip everything else
         // (the diagnostics will be collected during compilation)
-        if (await moduleDispatcher.RestoreModules(moduleDispatcher.GetValidModuleReferences(sourceFileGrouping.GetArtifactsToRestore()), forceRestore: forceRestore))
+        var artifactsToRestore = moduleDispatcher.GetValidModuleReferences(
+            sourceFileGrouping.GetExplicitArtifactsToRestore()).Concat(
+                sourceFileGrouping.GetImplicitArtifactsToRestore());
+
+        if (await moduleDispatcher.RestoreModules(artifactsToRestore, forceRestore: forceRestore))
         {
             // modules had to be restored - recompile
-            sourceFileGrouping = SourceFileGroupingBuilder.Rebuild(featureProviderFactory, moduleDispatcher, workspace, sourceFileGrouping);
+            sourceFileGrouping = SourceFileGroupingBuilder.Rebuild(featureProviderFactory, moduleDispatcher, ConfigurationManager, workspace, sourceFileGrouping);
         }
-
         return Create(sourceFileGrouping);
     }
 
-    public async Task<ImmutableDictionary<BicepSourceFile, ImmutableArray<IDiagnostic>>> Restore(Compilation compilation, bool force)
+    public async Task<ImmutableDictionary<BicepSourceFile, ImmutableArray<IDiagnostic>>> Restore(Compilation compilation, bool forceArtifactRestore)
     {
         var workspace = new Workspace();
         var sourceFileGrouping = compilation.SourceFileGrouping;
-        var originalModulesToRestore = sourceFileGrouping.GetArtifactsToRestore().ToImmutableHashSet();
-        if (await moduleDispatcher.RestoreModules(moduleDispatcher.GetValidModuleReferences(originalModulesToRestore), force))
+        var originalModulesToRestore = sourceFileGrouping.GetExplicitArtifactsToRestore().ToImmutableHashSet();
```

Comment:
Why don't we want to restore implicit artifacts here too?

Reply:
We do, good point!